Authentication context propagation

jdost opened the pull request:
Adds wrappers and handlers for retrieving propagated authentication information from upstream requests into an easy to use interface and standard system for further downstream propagation.  Overall the process is to convert the JWT payload received from the authentication service into an `AuthenticationContext` which can do validation and extraction, then store this is the current context and downstream requests via the built in thrift context will pull this value out of the current context and store the necessary information in a standard way for downstream services to easily and automatically retrieve.

There are some assumptions that may need to be refined, such as the way that the public RSA key for the authentication service is stored and retrieved, currently it is assumed (and hardcoded) to be in a `SecretStore` object under the simple key of `jwt/authentication/secret`.  Also, it is currently dependent on the originating service of the authentication to correctly package and define the information in the ongoing context for retrieval in the context handlers.  Any HTTP requests to downstream services will require the correct header be defined by hand (as there is no baseplate based HTTP client).

jdost commented on baseplate/core.py:
```diff
@@ -123,6 +124,123 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        if not self.secrets:
+            raise UndefinedSecretsException
+
+        return self.secrets.get_simple("jwt/authentication/secret")
```
This could have some more defined error checking, currently the error that the `SecretsStore` throws if the simple key lookup fails is going to pass through here uncaught, which may be preferable or may cause issues.

spladug replied:
Seems reasonable to allow the original exception up here. It's a fatal condition.

jdost commented on baseplate/integration/pyramid.py:
```diff
@@ -69,9 +77,12 @@ class BaseplateConfigurator(object):
         self.baseplate = baseplate
         self.trust_trace_headers = trust_trace_headers
+        self.auth_factory = auth_factory if auth_factory \
+            else AuthenticationContextFactory()
```
This is mostly to gracefully handle any downstream services that are not setup to parse and handle authentication tokens from having issues propagating the headers further.  If this conditional is reached, attempting to read the token will cause errors due to undefined secrets in the handler.

jdost replied:
Yeah that is probably better, I was not sure if I wanted to do more for the conditional check on the param but that is probably YAGNI...

jdost commented:
Also, this currently will fail for `python3` tests as we do not have a `python3-cryptography` package, which the `pyjwt` plugin requires to do the RSA based tokens.

pacejackson commented on baseplate/core.py:
```diff
@@ -123,6 +124,123 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        if self.token is None:
+            raise UndefinedAuthenticationError
+        if not self.valid:
+            return None
```
Relying on making these calls before accessing `self.payload` seems like a bug waiting to happen.  Maybe it makes more sense to have a `payload` `@property` method that checks these?

```python
@property
def payload(self):
    if self.token is None:
        raise UndefinedAuthenticationError
    if not self.valid:
        return {}
    return self._payload
```

jdost replied:
Originally I was wrapping the payload access entirely to allow for the error-ing (and had the no token error raise in the `valid` method) but wanted to make the conditions for this error to be more granular.  Only the `sub` field of the payload is being tested here as opposed to other (future) fields that may come down the line which I don't know if they should throw errors until we define how they should be handled in a richer format (scopes and origin service have both been discussed) and I would probably think those would have different conditions for errors.  I am kind of torn with this being adaptable for future proofing versus this being strict to avoid contextual missteps in the future.

pacejackson replied:
I waffled back and forth on this too while I was thinking about the comment.  Maybe just renaming `self.payload` to `self._payload` is a nice medium where it communicates to developers that they shouldn't be messing with it and we can leave the checks the way they are?

jdost replied:
Yeah, I am not super familiar with how much we would dislike it if new fields were being added to the token and applications would need to stall for smart wrappers to access the new fields.  I can see doing what you had but not raise the `UndefinedAuthenticationError` to at least encapsulate the validation check but not attempt to do any more, it would also enforce the `payload` property to be read only, which is a plus.

pacejackson commented on docs/baseplate/core.rst:
```diff
@@ -58,6 +58,37 @@ integrations <integration/index>`.
+
+When receiving an initial authentication token from the authentication service, the
+application should properly store the token in an
+:py:class:`~baseplate.core.AuthenticationContext` object attached to the current
```
Should we add a helper function to do this so developers don't have to worry about setting it to the exact variable name that we want?

```python
def attach_authentication_context(context, authentication_context):
    context.authentication = authentication_context
```

jdost replied:
I had thought about adding some classmethod to do the construction and binding, but also wanted to give something more explicit for how the binding looks.  I will ponder this some more...

spladug replied:
I like the way you've done this with the `attach_context` method.

pacejackson replied:
agreed

pacejackson commented on tests/integration/pyramid_tests.py:
```diff
@@ -51,6 +60,22 @@ def setUp(self):
+            }
+        }
+        secrets = store.SecretsStore("/secrets")
+        secrets._filewatcher = mock_filewatcher
```
Rather than mocking out `secrets._filewatcher` it seems cleaner to me to mock out the SecretStore interface.  `AuthenticationContext` cares about the SecretStore but mocking out the FileWatcher makes these tests concerned about the inner workings of the SecretStore object rather than the interface it provides.

jdost replied:
I just copied this pattern from how the `get_simple` method on the secrets test was setup and run.  I didn't see any pattern for test helpers, but was going to try and consolidate secret creation for tests that need it into something shared rather than this ugly copy pasta that I used here.

jdost commented on baseplate/core.py:
```diff
@@ -123,6 +124,123 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
         return cls(trace_id, parent_id, span_id, sampled, flags)
 
 
+class AuthenticationContextFactory(object):
```
I was afraid of the misunderstanding of what `Authentication` would mean for this.  The token conveys only a basic sense of authentication and the existence of the object does not necessarily correlate to having authentication, it just encapsulates the actual checking of the current authentication.  I don't know if the actual names `AuthenticationContext{,Factory}` make sense but I would prefer that whatever we call it, the wrapper and factory share the same basis rather than have an `[A]Factory` produce not an `[A]`...

spladug replied:
:dizzy_face: I think I've ended up thinking this is good as is but totally see the confusion potential of `ContextFactory`

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        return self._valid
+
+    @property
+    def payload(self):
```
Perhaps we should just not allow this so that people can't rely on internals of the auth token?

jdost replied:
As we discussed before I made this change, I am fearful of new fields being added to to the token in the authentication service response and it requiring applications wait for a new baseplate release and upgrade in order to access them.  Ideally all of the attributes will have intelligent property wrappers to access them, but having a hard limitation feels more inhibiting for developers than beneficial.

spladug replied:
On the flip side, I'm worried that applications would end up tied to the implementation of the token and it'd be really hard to change.

jdost replied:
The optimist in me would like to think that it would make it to baseplate and be implemented using the methods there sooner rather than later...

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        if self.token is None:
+            raise UndefinedAuthenticationError
+
+        return self.payload.get("sub", None)
```
This seems dangerous since an application could get the account_id without ever having checked validity?

jdost replied:
The `self.payload` property calls it, so it will raise it there.  If we want to not expose the payload itself, then the check would move back into here.

spladug replied:
Ah! Right, sorry. I got confused by `payload` vs `token`; kinda not dissimilar conceptually in this context.

jdost replied:
Yeah, I actually used the wrong one in a few places during development.  I have thought about using something like `decrypted_token` but don't really like using such juicy words as `decrypted` in attribute names.  Ideally nothing outside of the wrapper accesses these attributes but we all know there will be some situation where it will make sense...

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+
+        return self.payload.get("sub", None)
+
+    def __str__(self):
```
Maybe a `__repr__` instead that just shows some tidbits about the object? I worry about putting user info into logs too much.

jdost replied:
I am going to drop this, it was meant to be a wrapper for logic when "casting" the object into the header field, but too much logic was living outside of this to give it any real value.

spladug commented on docs/baseplate/core.rst:
```diff
@@ -58,6 +58,38 @@ integrations <integration/index>`.
+
+    authentication_token = retrieve_token_from_authentication_service()
+    authentication_context = AuthenticationContext(authentication_token, secrets_store)
+    authentcation_context.attach_context(context)
```
`authentcation`

spladug replied:
(could turn this into a doctest if you want to be sure it works now and in the future)

jdost replied:
The problem is that I have a "magical" method for showing token retrieval and undefined `secrets_store` and `context` variables used... :(

spladug replied:
You can define a mock method in the test setup block.

spladug commented on setup.py:
```diff
@@ -8,6 +8,7 @@
 install_requires = [
     "requests",
     "posix_ipc",
+    "pyjwt",
```
just to confirm: this ends up in the debian package dependencies right?

you'll probably also need to update `docs/requirements.txt` for doc building to work right still (shitty, sorry, i know :()

jdost replied:
This would also need to add some other packages (`cryptography`) and I would need to confirm it works with the debian package builder...

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+    """
+    def __init__(self, token=None, secrets=None):
+        self.token = token
+        self.secrets = secrets
```
can this be a private as well? no need for downstream users to know we've got a copy of the secrets store around (they should be using it from their own contexts for law of demeter etc.)

jdost replied:
Probably, I left it accessible because it doesn't feel too scary to expose and I can't see much of an argument about what exposing it could hurt, I may spend some time and rework this and try to refactor how the location of the public key is defined (rather than having it hard-coded as it is now).

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+class AuthenticationContext(object):
+    """Wrapper for the contextual authentication information
+
+    :param str token: the JWT value for the currently propagated authentication
```
There are a few places in the docs that explicitly mention that the token is a JWT. I think it'd be safer to continue describing it as an opaque authentication token as the clients should never be touching the JWT envelope themselves.

spladug replied:
Maybe just don't bother documenting the parameters on this class at all since the (internal) factory should be the only one instantiating it?

jdost replied:
Not always.  For the originating service, it would explicitly calling to the authentication service and would be handling the raw payload from there.  We could write up helpers to handle talking to this helper which would handle this, but until we do, the token needs to be touched in some way.

spladug replied:
The token as an object does, yes, but not its guts.

jdost replied:
The comment was mostly in reference to your second comment (about not even documenting because only the factory is building it).  It is also a concern for another subsequent comment about double defaulting.

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, token=None, secrets=None):
```
Why defaults here if the factory's instantiating it?

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+            self._payload = jwt.decode(self.token, self._secret(),
+                                       algorithms='RS256')
+            self._valid = True
+        except jwt.ExpiredSignatureError:
```
Looks like they all inherit from `InvalidTokenError` if you wanna reduce branches here: https://github.com/jpadilla/pyjwt/blob/master/jwt/exceptions.py

jdost replied:
I was not *super* confident we wanted to eat all of them.  I was trying to get an idea of various failure scenarios and try to only explicitly catch them (versus something like the token getting corrupted or other wicked things) and wanted to only handle the things I had seen (even though `DecodeError` shouldn't happen for correctly designed applications, I saw it when trying to use very short RSA keys).

spladug replied:
Ahh interesting, that's a reasonable argument. Maybe comment that others are explicitly uncaught and why?

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        """
+        if self._valid is not None:
+            return self._valid
+        elif self.token is None:
```
I feel like this should either just return `False` or raise an exception. It's kinda weird to have a three-state boolean.

jdost replied:
I didn't know how we would like to handle complete unauthenticated calls.  Because in that ideal, the authentication is not really valid or invalid, it just doesn't exist.  I am unsure if it is guaranteed that all calls into a service will have authentication attached and how services should handle if they are not, but a lack of authentication sounds like a different failure scenario than an expired authentication.

spladug replied:
Is there any time you'd want to call `valid` and distinguish those though?

jdost replied:
I could see having another attribute on the object that is just `self.defined = self.token is not None` and using that instead and then just having this be
```
elif not self.defined:
    return False
```
🤔

spladug commented:
> Also, this currently will fail for python3 tests as we do not have a python3-cryptography package, which the pyjwt plugin requires to do the RSA based tokens.

Can you mark the relevant test `xfail` when PY3 please? Then the suite will pass but call out the expected failure.

pacejackson commented:
Do we want PY3 to fail though? We currently have python 3 baseplate services running in prod.

spladug commented on baseplate/context/thrift.py:
```diff
@@ -104,6 +104,12 @@ def _call_thrift_method(self, name, *args, **kwargs):
+                                and span.context.authentication.token:
+                            prot.trans.set_header(
+                                "Authentication",
+                                str(span.context.authentication)
```
It feels a bit weird to explicitly check if `.token` is available and then use `__str__` here. Maybe do one or the other? (I lean towards just using `.token` explicitly both places)

jdost replied:
That makes sense.  Originally I was trying to have the token conversion work via the string casting of the object, but could not get it to handle unset tokens cleanly.  I was hoping to have the case of an unset token handled internal to the wrapper rather than here so that the hand rolled HTTP implementations would not be too hairy.

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
 
 
+class AuthenticationContextFactory(object):
+    """Building factory for :py:class:`AuthenticationContext` objects, handles
```
"building factory"?

pep-257 nit: the first line of a docstring should be short enough to fit on one line.

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line.

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        self.secrets = secrets_store
+
+    def make_context(self, token):
+        """:py:class:`AuthenticationContext` builder
```
pep-257 nit:

> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

jdost commented:
I have a stashed change to the `.travis.yml` file that adds all of the package dependencies, so I can add this but comment out the missing package and update the tests.  Will do it in a bit...

Python3 itself will not fail, it just doesn't have the package dependencies built (you can still install using `pip`) to have the travis instance set up in the standard way we are using (`python3-*` debian packages).

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+        return self.token if self.token else ""
+
+
+class UndefinedSecretsException(Exception):
```
"undefined" reads a bit weirdly to me here: how about `AuthenticationSecretsNotConfiguredException` ? (very long... hrm)

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+    """
+    def __init__(self):
+        super(UndefinedSecretsException, self).__init__(
+            "No SecretsStore defined.")
```
might want to give a specific "defined for authentication context" or something. these exception strings sometimes end up devoid of context when doing single-line logging.

jdost replied:
yeah, I always envisioned this error being pretty show stopping and mostly seen on startup or in development.

spladug commented on baseplate/core.py:
```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
+            "No SecretsStore defined.")
+
+
+class UndefinedAuthenticationError(Exception):
```
"undefined" again: how about `NoAuthenticationContextError` ?

jdost replied:
Agree it is not a good name, but `NoAuthenticationContextError` being raised *from* a method on an `AuthenticationContext` object seems super confusing.  I will ponder this name and see if I can find a good one.

spladug replied:
Agh, you're right, that's way more confusing! `NoAuthenticationTokenError`?

spladug commented on docs/baseplate/core.rst:
```diff
@@ -58,6 +58,38 @@ integrations <integration/index>`.
+Authentication
+--------------
+
+If the application is dealing with authentication either directly from the
```
"dealing with authentication" maybe call out why you would want to deal with authn, like "if the application wants to check the authenticated user for a given request to authorize data access..."

spladug commented:
@pacejackson py3 wouldn't fail, it just wouldn't be able to validate authn tokens until we update cryptography. Hoping to punt that work until the self-serve packaging stuff happens later this quarter and lets us skip a debianization step :)

jdost commented on tests/unit/core_tests.py:
```diff
@@ -14,11 +15,21 @@
+try:
+    import cryptography
+except:
+    cryptography_installed = False
```
This is ugly and I don't like it very much, but that fact may be a feature rather than a bug and will incentivize me to clean this up sooner rather than later.

jdost commented on tests/unit/core_tests.py:
```diff
@@ -226,3 +237,55 @@ def test_from_upstream_handles_no_sampled_or_flags(self):
+class AuthenticationContextTests(unittest.TestCase):
+    VALID_TOKEN = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X3VzZXJfaWQiLCJleHAiOjQ2NTY1OTM0NTV9.Q8bz2qccFOHLTQ6H3MPdjSh7wDkRQtbBuBwGMzNRKjDFSkCoVF5kiwhBUdwbW8UXO5iZn4Bh7oKdj69lIEOATUxFBblU8Do05EfjECXLYGdbr6ClNmldrB8SsdAtQYQ4Ud-70Z8_75QvkqX_TY5OA4asGJZwH9MC7oHey47-38I"
+    EXPIRED_TOKEN = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0X3VzZXJfaWQiLCJleHAiOjE1MDI5OTM3NTN9.OPBIxaOEx0hnnB_wrfCuqfSIeP0a1abNdoZ2KejXReKeETQathr-PW2GqAhjGcUdCG3rXK8ezFKXdlB65kloqNdQii5b3qaJ5PDIdMNxY0Oi7TAqH86oog_umm7G-_p4MPPhRjxUm6Qp85-EaJUgyv26BUKSYY7-KyySjnrmP8g"
+    TOKEN_SECRET = "-----BEGIN PUBLIC KEY-----\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC0Kd3qYtc6zI5tj3iKBux70BhE\nZLLJ7fAKNBUO7h9FCwUcYku+SFigzNOu3AAYt3seNgxl+cvMR2+SNwsa605J9D1v\n9eGmpcITQi85SeJnfR7LJUMu7RieY5wEl0RyuwnSkX3Gkv0+hZISC/XYcWEYolIi\n8725u7u/8HRtUeHoLwIDAQAB\n-----END PUBLIC KEY-----"
```
Just to note, these values were generated using a private key I spun up locally (and this is the public key) but it is not used anywhere (it is a super insecure 1024 no-passphrase keypair).  The "valid" token has a TTL of about 100 years (if we are still here when this expires, I will have no issue knowing that we have become immortal) and the expired token expired like 1 second after I generated it.

jdost commented:
:haircut:

Cleaned up some naming to try and convey the overall scenario instead of literal issues, removed most of the references to tokens being JWT outside of the very internal decoding logic, cleaned up some of the testing to allow for CI to pass, other misc things...

jdost commented:
`UndefinedSecrets` makes sense to me only in that if it comes up, it will be in initial rollout or implementation/development.  So it is not super helpful in terms of execution, but should not get through to execution unless someone is deploying untested code.

`WithheldAuthentication` is not my favorite, but without making some long name (`NotProvidedAuthentication` was what I was going with but `withheld` is similar in meaning to `not provided` which is how I got there) I am unsure of better options...

spladug commented:
Yeah, `UndefinedSecrets` is probably fine.

I think my qualm with `Withheld` is that it implies someone (or something) deliberately didn't provide something it had. What about `MissingAuthentication`?
